Propagate listener task exceptions to the main loop.

fselmo opened the pull request:
### What was wrong?

Closes #3375 (along with #3387)

### How was it fixed?

As I understand it, the way asyncio works is each task operates under its own context. Directly communicating with another context and trying to immediately propagate an exception to the main loop, for example, does not seem to be a very straightforward thing to do surprisingly. For better or worse, I think when we poll for messages we should check if the listener task is done and, if it is, if an exception was set. If an exception was set then we can raise it in the main loop where we are polling for messages.

- When we poll for messages, poll to see if the background task is finished. If it is, handle any exceptions in the main loop. On subscription streams, since we hang awaiting a message to appear in the subscription queue, we need to signal when the background task is `done` since no messages will be coming in and we will hang indefinitely. Push a new exception `TaskNotRunning` to the queue. Revamp the queue to raise when this is pushed to it. Handle this exception in the subscription polling.
- Refactor some of the logic. When disconnecting persistent connection providers, shut down the listener task before closing the connection to avoid any sort of errors when the task is still trying to read from the closing connection.

---

I added some nice logging that really shouldn't be too noisy unless the connection is reconnecting all the time (which I forced for the example here).

```
INFO:web3.providers.WebSocketProvider:Connecting to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:Successfully connected to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:WebSocketProvider listener background task started. Storing all messages in appropriate request processor queues / caches to be processed.
newHeads sub_id: 0xad65c6e2aefa417dafda6999d0bd774b
INFO:web3.providers.WebSocketProvider:Subscription response queue synced with websocket message stream.
blocknum: 19829151
blocknum: 19829152
ERROR:web3.providers.WebSocketProvider:Connection interrupted, attempting to reconnect...
INFO:web3.providers.WebSocketProvider:Message listener background task successfully shut down.
INFO:web3.providers.WebSocketProvider:Successfully disconnected from: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:Connecting to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:Successfully connected to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:WebSocketProvider listener background task started. Storing all messages in appropriate request processor queues / caches to be processed.
newHeads sub_id: 0x718852a668ed437fbdc86243db929405
blocknum: 19829153
blocknum: 19829154
ERROR:web3.providers.WebSocketProvider:Connection interrupted, attempting to reconnect...
INFO:web3.providers.WebSocketProvider:Message listener background task successfully shut down.
INFO:web3.providers.WebSocketProvider:Successfully disconnected from: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:Connecting to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:Successfully connected to: wss://mainnet.infura.io/ws/v3/c364...
INFO:web3.providers.WebSocketProvider:WebSocketProvider listener background task started. Storing all messages in appropriate request processor queues / caches to be processed.
newHeads sub_id: 0xf46690d2853c4afba527ac71dd1915e0
blocknum: 19829155
...
```
---
Bonus:

I've been meaning to do this for some time and I think this is an appropriate time. Refactor the commonality of the listener task down to the base `PersistentConnectionProvider` class and define provider-specific methods that either need to be implemented or can be overridden to fine-tune logic specific to each persistent connection provider. This keeps things quite a bit DRYer.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![31205](https://github.com/ethereum/web3.py/assets/3532824/58c5dd6e-13eb-4b18-9203-6c3acda37936)

fselmo committed:
Propagate listener task exceptions to the main loop.

- Raise exceptions in the message listener task if not silenced. Check
  if the task is completed and if an exception was set when polling for
  results in the caches and raise the exception if it was set.

- Refactor some of the logic. When disconnecting persistent connection
  providers, shut down the listener task before closing the connection
  to avoid any sort of errors when the task is still trying to read from
  the closing connection.

fselmo commented:
I'm actually going to re-structure this a bit. Will ping for review when ready again.

fselmo committed:
Refactor common logic to persistent connection provider base class:

- Define a base pattern for the listener task on the base
``PersistentConnectionProvider`` class. Define provider-specific
methods that can be configured on the implementation classes in order
to handle the provider-specific logic and error logging.

fselmo commented:
Just kidding this is still in progress 😔

fselmo committed:
Properly handle checking sub messages while checking listener task:

- Simultaneously check for subscription messages while checking
  if the listener task is done, raising any exceptions if they occurred
  in the listener task.

- Add tests for subscriptions with iterator pattern (would've failed before
  this commit).

- This adds a bit of added overhead for the subscription message stream.
  Change INFO logs to DEBUG when the message stream is out of sync with
  the websocket connection. These aren't super useful to the end user
  and can be noisy now.

fselmo committed:
Move more shared logic to base class for persistent connection providers

fselmo committed:
Make polling the sub queue and checking listener task more performant:

- Instead of using ``asyncio.wait()`` to poll which task finishes
  first, simply push a ``None`` value to the sub queue as part of
  the callback when the listener task finishes. This tells any
  message stream that the listener task has finished, pops the
  iterator out of polling the queue, and allows it to address any
  listener task exceptions that may have occurred.

fselmo commented:
@kclowes @pacrob @reedsa Hey it's me again. This is ready now 😄. We were missing a good test for subscription polling with the iterator pattern. This would've failed or hanged forever. That should be addressed now.

fselmo committed:
Add some useful logging around the persistent provider connection.

kclowes commented on web3/_utils/module_testing/module_testing_utils.py:
```diff
@@ -189,6 +195,13 @@ async def __anext__(self) -> bytes:
         return self.messages.popleft()
 
+    @staticmethod
+    async def pong() -> bool:
```
nit: 
```suggestion
    async def pong() -> False:
```

kclowes commented on web3/providers/persistent/websocket.py:
```diff
@@ -116,47 +115,13 @@ async def is_connected(self, show_traceback: bool = False) -> bool:
-            await self._message_listener_task
-        except (asyncio.CancelledError, StopAsyncIteration):
-            pass
-        self._request_processor.clear_caches()
```
I think we lost this line in the persistent provider's `disconnect`. Was that on purpose?

fselmo replied:
Nope, this should be there! Thanks for catching. I'll see if I can fit this into a test too.

kclowes commented on web3/main.py:
```diff
@@ -511,12 +512,11 @@ async def __aexit__(
         "when instantiating via ``async for``."
     )
     async def __aiter__(self) -> AsyncIterator[Self]:
-        if not await self.provider.is_connected():
```
Does this check matter? I can't decide, ha

fselmo replied:
I think we should try to keep `connect` where it re-creates the state each time and force a re-connect on each iteration to make sure we are creating a clean slate. Thoughts there?

fselmo replied:
Also, since we're disconnecting after yielding now, I think this makes more sense.

kclowes replied:
okay, cool. That makes sense to me!

kclowes reviewed (approved):
Nice tests and nice refactor! This looks good to me overall. Pushing `None` to the queue is the only thing that stands out as something that might be a little foot-gun-y. In my head, it seems like it would be more descriptive and maybe less error prone if we could raise a custom exception there, but I can't quite see the path to how that could work. So I'm good leaving as-is, but wanted to flag in case you saw a nice solution off the bat.

fselmo commented:
> Pushing None to the queue is the only thing that stands out as something that might be a little foot-gun-y.

I think you're absolutely right. I don't see `None` being pushed to the cache in another way but I think we can isolate this case to be more unique. If I can't, I think we can still use `None` to un-brick the awaiting of a message and we can make sure we make the proper checks of whether or not the task really did finish and if it didn't we just move on to the next iteration of the subscription messages - essentially causing a no-op in that scenario.

fselmo committed:
Use a ``TaskReliantQueue`` to handle when a reliant task is not running:

- More gracefully and explicitly handle when a queue relying on a task
  that is not running is being awaited on for a result. This involves
  raising a new ``TaskNotRunning`` exception when a task is not running
  which is pushed to the queue in the reliant task's completion callback.

fselmo committed:
address comments from PR #3378: clear caches + add tests

fselmo commented:
@kclowes I came up with a more elegant solution for the queue to operate correctly as long as the listener task is running

> In my head, it seems like it would be more descriptive and maybe less error prone if we could raise a custom exception there

taking this as inspiration, lmk what you think: a4790d6

...other comments addressed in 8ed6df9

kclowes commented:
> I came up with a more elegant solution for the queue to operate correctly as long as the listener task is running

super elegant 🤩  🚀
